src/cargo/ops/common_for_install_and_uninstall.rs

```diff
@@ -544,12 +544,33 @@ where
             let pkg = Box::new(source).download_now(pkgid, config)?;
             Ok(pkg)
         }
-        None => bail!(
-            "could not find `{}` in {} with version `{}`",
-            dep.package_name(),
-            source.source_id(),
-            dep.version_req(),
-        ),
+        None => {
+            let version: String = dep.version_req().to_string();
+            match PackageId::new(dep.package_name(), &version[1..], source.source_id()) {
+                Ok(pkg_id) => {
+                    if source.is_yanked(pkg_id).unwrap_or(false) {
+                        bail!(
+                            "cannot install package `{}`, it has been yanked from {}",
+                            pkg_id.name(),
+                            pkg_id.source_id()
+                        )
+                    } else {
+                        bail!(
+                            "could not find `{}` in {} with version `{}`",
+                            dep.package_name(),
+                            source.source_id(),
+                            dep.version_req(),
+                        )
+                    }
+                }
+                Err(_) => bail!(
+                    "could not find `{}` in {} with version `{}`",
```

Comment:
Can this be restructured to avoid the duplication of these two cases?  Maybe something like:

```rust
let maybe_pkg_id = PackageId::new(dep.package_name(), &version[1..], source.source_id());
let is_yanked = maybe_pkg_id.map_or(false, |pkg_id| source.is_yanked(pkg_id).unwrap_or(false));
if is_yanked {
    // ...yanked error
} else {
    // ...not found error
}
```

Reply:
Yes, you are correct. Thanks for the snippet. I've incorporated this in the PR. Please have a look once you get time.